File: samples/android/app/src/main/java/com/example/zumoquickstart/TodoItemAdapter.java
package com.example.zumoquickstart;

import android.app.Activity;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;

public class TodoItemAdapter  extends ArrayAdapter<TodoItem> {
    /**
     * Adapter context
     */
    Context mContext;

    /**
     * Adapter View layout
     */
    int mLayoutResourceId;

    public TodoItemAdapter(Context context, int layoutResourceId) {
        super(context, layoutResourceId);
        mContext = context;
        mLayoutResourceId = layoutResourceId;
    }

    /**
     * Returns the view for a specific item on the list
     */
    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        View row = convertView;
        TodoItem currentItem = getItem(position);

        if (row == null) {
            LayoutInflater inflater = ((Activity) mContext).getLayoutInflater();
            row = inflater.inflate(mLayoutResourceId, parent, false);
        }

        row.setTag(currentItem);
        CheckBox checkBox = (CheckBox) row.findViewById(R.id.checkToDoItem);
        checkBox.setText(currentItem.getText());
        checkBox.setChecked(false);
        checkBox.setEnabled(true);
        checkBox.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (checkBox.isChecked()) {
                    checkBox.setEnabled(false);
                    if (mContext instanceof TodoActivity) {
                        ((TodoActivity) mContext).checkItem(currentItem);
                    }
                }
            }
        });

        return row;
    }
}

File: samples/android/app/src/main/java/com/azure/mobile/zumoquickstart/TodoItemAdapter.kt
package com.azure.mobile.zumoquickstart

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView

class TodoItemAdapter(val onChangeListener: (TodoItem, Boolean) -> Unit)
    : ListAdapter<TodoItem, TodoItemAdapter.ViewHolder>(DiffCallback())
{
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TodoItemAdapter.ViewHolder
        = ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.row_list_item, parent, false))

    override fun onBindViewHolder(holder: TodoItemAdapter.ViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    /**
     * Part of the ListAdapter specification - this class helps with determining if the items
     * are the same or not.
     */
    private class DiffCallback: DiffUtil.ItemCallback<TodoItem>()
    {
        override fun areItemsTheSame(oldItem: TodoItem, newItem: TodoItem): Boolean
            = oldItem.id == newItem.id

        override fun areContentsTheSame(oldItem: TodoItem, newItem: TodoItem): Boolean
            = (oldItem.text == newItem.text) && (oldItem.complete == newItem.complete)
    }

    /**
     * Binding for each row.
     */
    inner class ViewHolder(private val containerView: View) : RecyclerView.ViewHolder(containerView)
    {
        fun bind(item: TodoItem) {
            val checkbox = containerView.findViewById<CheckBox>(R.id.check_todo_item)

            checkbox.text = item.text
            checkbox.isChecked = item.complete
            checkbox.setOnCheckedChangeListener { _, isChecked ->
                onChangeListener?.invoke(item, isChecked)
            }
        }
    }
}